### PR TITLE
Google Ad Manager Feature Extraction

### PR DESCRIPTION
**Pull request by jeffersonrabb**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR describes the extraction of the Google Ad Manager functionality to a feature plugin. To keep things simple I'll describe the complete change here even though it technically spans three repositories. 

This work establishes a pattern for extracting Newspack functionality to feature plugins. The approach:

- Move all new functionality to a feature plugin (`newspack-{feature-description}`)
- Create a Configuration Manager class to interact with the plugin.
- Try to limit all Newspack Plugin Wizard activity to API endpoints.

The elements of this PR are:

- Created a feature plugin (https://github.com/Automattic/newspack-gam) to contain Newspack Google Ad Manager features.
- Moved the Ad Manager block from https://github.com/Automattic/newspack-blocks/pull/39 to the new plugin. When this lands, https://github.com/Automattic/newspack-blocks/pull/39 can be abandoned and closed.
- Moved all of the Ad Unit model code from `newspack-plugin` to `newspack-gam` (https://github.com/Automattic/newspack-gam/blob/master/includes/class-newspack-gam-model.php), including [unit tests](https://github.com/Automattic/newspack-gam/tree/master/tests/unit-tests). The `newspack-plugin` [Wizard's PHP class](https://github.com/Automattic/newspack-plugin/compare/update/use-gam-plugin?expand=1#diff-7037d54febdb3a09e7453fc8eb4516d7) is dramatically simpler, containing only API endpoints and script/style enqueueing. 
- Created a [Configuration Manager class](https://github.com/Automattic/newspack-plugin/compare/update/use-gam-plugin?expand=1#diff-0af9a1b981f2f2343689b108042cf35f) for `newspack-gam`.  All interaction with the feature plugin occurs in this class. At first glance this class may seem unnecessary, but I like adhering to the Wizard->Configuration Manager->Plugin model even when the plugin is our own.
- The [Google Ad Manager block](https://github.com/Automattic/newspack-gam/tree/master/src/blocks/google-ad-manager) addresses feedback from https://github.com/Automattic/newspack-gam/issues/1. 
- https://github.com/Automattic/newspack-plugin/blob/master/includes/wizards/class-google-ad-manager-wizard.php#L141-L150 was removed. If the AMP plugin detects an `amp-ad` element, it will automatically add this script.
- Declared  `newspack-gam` as a required plugin for the Ad Manager Wizard. 

### How to test the changes in this Pull Request:

1. Apply this PR to the Newspack plugin and run `npm run build`
1. Clone https://github.com/Automattic/newspack-gam in your test environment's plugins directory. Run `composer install`. Run `npm install && npm run build:webpack`. After this branch lands, I'll generate a release of `newspack-gam` and add the path to the .zip to `newspack-plugin`.
1. Navigate to the Ad Manager Wizard. This should install `newspack-gam` before opening.
1. Create/Edit/Delete ad units. Sample AMP code can be found in the description of https://github.com/Automattic/newspack-blocks/pull/39
1. Create a post and add the Google Ad Manager block (found in the Newspack category).
1. Verify the dropdown lists all Ad Units you have set up.
1. If you have set up no Ad Units verify the block shows UI with text and a link to create new Ad Units.
1. Verify the block placeholder dimensions change to match the Ad Unit you have selected.
1. Verify that choosing an Ad Unit that does not contain proper dimension data displays a block error notification. Also verify that switching to a valid Ad Unit clears the notification.
1. Verify that all unit tests pass in `newspack-gam`
1. Disable `Newspack Blocks` and verify that the `Google Ad Manager` block is still available, in the Common category.  

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-plugin-manager.php [hunk 1]
```diff
@@ -23,11 +23,18 @@ class Plugin_Manager {
 	 *
 	 * @return array of plugins info.
 	 */
-
 	public static function get_managed_plugins() {
 		$managed_plugins = [
 			'newspack-blocks'               => [
 				'Name'        => esc_html__( 'Newspack Blocks', 'newspack' ),
+				'Description' => esc_html__( 'Google Ad Manager integration.' ),
```
Comment (philipjohn) [inline includes/class-plugin-manager.php]: Looks like the descriptions have ended up the wrong way round here :)

File: includes/configuration_managers/class-newspack-gam-configuration-manager.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		return $this->is_configured() ?
+			\Newspack_GAM_Model::get_ad_units() :
+			$this->unconfigured_error();
+	}
+
+	/**
+	 * Get a single ad unit.
+	 *
+	 * @param number $id The id of the ad unit to retrieve.
+	 */
```
Comment (philipjohn) [inline includes/configuration_managers/class-newspack-gam-configuration-manager.php]: Minor: It'd be good to have the `@return` documented in these methods too.

**Comment by jeffersonrabb** [comment]
> I'm currently not seeing any ads on the front end. Here's a couple of ad codes I'm using, for example:

Discussed this with @philipjohn and we determined that this was due to branch not being up-to-date on the test instance. After updating the problem was resolved.

**Comment by jeffersonrabb** [comment]
> Another issue I'm seeing is that when I use an AMP Ad, the amp-ad component JS is not automatically added by the AMP plugin, which I didn't think the AMP plugin actually did anyway, but maybe I'm assuming wrong.

The AMP plugin does seem to include the `amp-ad` script if it encounters any `amp-ad` elements. For my testing of this I've been using AMP Transitional mode to allow switching between the environments. When viewing an AMP page with `<amp-ad/>`s in it, the script is being enqueued.

File: includes/class-plugin-manager.php (hunk 1 above)
Comment (jeffersonrabb) [reply includes/class-plugin-manager.php]: Good catch: https://github.com/Automattic/newspack-plugin/pull/184/commits/d6803795783ee4a155ec303b1055d70d58a19b2b

File: includes/configuration_managers/class-newspack-gam-configuration-manager.php (hunk 2 above)
Comment (jeffersonrabb) [reply includes/configuration_managers/class-newspack-gam-configuration-manager.php]: Yup, good call: https://github.com/Automattic/newspack-plugin/pull/184/commits/07062063fe08d9ef3779d158e91af3233447f9a9

**Comment by jeffersonrabb** [comment]
Thanks for the review!

**Pull request merged**